services/server/src/server/services/storageServices/AbstractDatabaseService.ts

```diff
@@ -0,0 +1,374 @@
+import {
+  Match,
+  CheckedContract,
+  Transformation,
+} from "@ethereum-sourcify/lib-sourcify";
+import { id as keccak256str } from "ethers";
+import * as Database from "../utils/database-util";
+import { Pool } from "pg";
+
+export default abstract class AbstractDatabaseService {
+  abstract init(): Promise<boolean>;
+  abstract databasePool: Pool;
+  abstract databaseName: string;
+
+  validateBeforeStoring(
+    recompiledContract: CheckedContract,
+    match: Match
+  ): boolean {
+    if (
+      recompiledContract.runtimeBytecode === undefined ||
+      recompiledContract.creationBytecode === undefined ||
+      match.onchainRuntimeBytecode === undefined ||
+      match.onchainCreationBytecode === undefined
+    ) {
+      throw new Error(
+        `can only store contracts with both runtimeBytecode and creationBytecode address=${match.address} chainId=${match.chainId}`
+      );
+    }
+    if (match.creatorTxHash === undefined) {
+      throw new Error(
+        `can only store matches with creatorTxHash address=${match.address} chainId=${match.chainId}`
+      );
+    }
+    return true;
+  }
+
+  getKeccak256Bytecodes(recompiledContract: CheckedContract, match: Match) {
+    return {
+      keccak256OnchainCreationBytecode: keccak256str(
+        match.onchainCreationBytecode!
+      ),
+      keccak256OnchainRuntimeBytecode: keccak256str(
+        match.onchainRuntimeBytecode!
+      ),
+      keccak256RecompiledCreationBytecode: keccak256str(
+        recompiledContract.creationBytecode!
+      ),
+      keccak256RecompiledRuntimeBytecode: keccak256str(
+        recompiledContract.runtimeBytecode!
+      ),
+    };
+  }
+
+  async getDatabaseColumns(
```

Comment:
Instead of having a DatabaseColumns type that has all columns dumped, wouldn't it be neater to have each columns of each table in its designated type? This way we will have the data structure of the database in our typings. So like:
```ts
interface DatabaseCode {
  bytecodeHash: string,
  bytecode: string
}

...

interface DatabaseVerifiedContract {
  compilationId: string,
  contractId: string,
  creationTransformations: Transformation[],
  creationTransformationValues: TransformationValues,
  ...
}
```

Then we can prepare these types either in `insertNewVerifiedContract` or again in a preparation function similar to `getDatabaseColumns`

Reply:
Edit: You actually have the types in `database-utils.ts`. Why aren't using them and why a separate `DatabaseColumns` dump type?  Why don't you just return all entries to be written and just pass them inside the database-utils functions such as `insertCompiledContract(pool: Pool, compiledContract: Tables.CompiledContract)`?

Reply:
Try to look at how I changed it :)